mlir/include/mlir/Dialect/MIOpenOps/MIOpenCPP.h

```diff
@@ -91,51 +91,325 @@ T integer_least_multiple(T x, T y)
+      if (dimIndexVal["ci"].first == 3) {
+        input2GemmKVectorizable = true;
+      } else {
+        input2GemmKVectorizable = false;
+      }
+    } else if (opType == mlir::miopen::ConvOpType::Conv2DBwdDataOpType) {
+      // For output tensor.
+      // When K is the fastest changing dimension(3),
+      // gemmK dimension is vectorizable, gemmN is not, and vice versa.
+      // Vectorization width depending on length of K.
+      if (dimIndexVal["ko"].first == 3) {
+        input2GemmKVectorizable = true;
+      } else {
+        input2GemmKVectorizable = false;
+      }
+    }
+  }
+
+  static void
+  obtainNCHWVecLen(llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal,
+                   int64_t &vecLen) {
+    if (dimIndexVal["ni"].first == 3) {
+      vecLen = dimIndexVal["ni"].second;
+    } else if (dimIndexVal["ci"].first == 3) {
+      vecLen = dimIndexVal["ci"].second;
+    } else {
+      // Not vectorizable
+      // TODO(optimize): For 1x1, stride 1, pad 0 conv, vecLen is hi * wi
+      vecLen = 1;
+    }
+  }
+  static void
+  obtainNKHWVecLen(llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal,
+                   int64_t &vecLen) {
+    if (dimIndexVal["ko"].first == 3) {
+      vecLen = dimIndexVal["ko"].second;
+    } else if (dimIndexVal["ko"].first == 0) {
+      // dimKO is the lowest changing dimension, which means dimN/dimHo/dimWo
+      vecLen = dimIndexVal["no"].second * dimIndexVal["ho"].second *
+               dimIndexVal["wo"].second;
+    } else if (dimIndexVal["ko"].first == 1) {
+      // Ko's position is at 1, vectorization legnth is last two dimensions
+      if (dimIndexVal["no"].first == 0) {
+        vecLen = dimIndexVal["ho"].second * dimIndexVal["wo"].second;
+      } else if (dimIndexVal["ho"].first == 0) {
+        vecLen = dimIndexVal["no"].second * dimIndexVal["wo"].second;
+      } else {
+        vecLen = dimIndexVal["no"].second * dimIndexVal["ho"].second;
+      }
+    } else {
+      // K's position is 2, vectorization legnth is last dimension
+      if (dimIndexVal["no"].first == 3) {
+        vecLen = dimIndexVal["no"].second;
+      } else if (dimIndexVal["ho"].first == 3) {
+        vecLen = dimIndexVal["ho"].second;
+      } else {
+        vecLen = dimIndexVal["wo"].second;
+      }
+    }
+  }
+
+  static void
+  obtainInput2VecLen(mlir::miopen::ConvOpType opType,
+                     llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal,
+                     int64_t &vecLen) {
+    if (opType == mlir::miopen::ConvOpType::Conv2DOpType) {
+      obtainNCHWVecLen(dimIndexVal, vecLen);
+    } else if (opType == mlir::miopen::ConvOpType::Conv2DBwdDataOpType) {
+      obtainNKHWVecLen(dimIndexVal, vecLen);
+    }
+  }
+
+  static void
+  obtainOutputVecLen(mlir::miopen::ConvOpType opType,
+                     llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal,
+                     int64_t &vecLen) {
+    if (opType == mlir::miopen::ConvOpType::Conv2DOpType) {
```

Comment:
comment here needed to describe why we are only looking for a specific layout, but not other layouts.